---
title: "MaAsLin 3"
date: "Friday, July 5, 2024"
author:
 - name: William Nickols
   email: willnickols@g.harvard.edu
 - name: Jacob Nearing
   email: nearing@broadinstitute.org
 - name: Sagun Maharjan
   email: smaharjan@hsph.harvard.edu
output: html_document
vignette: >
    %\VignetteIndexEntry{Maaslin3}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

# MaAsLin 3 User Manual #

MaAsLin 3 is the next generation of MaAsLin (Microbiome Multivariable Association with Linear Models).

[MaAsLin3](http://huttenhower.sph.harvard.edu/MaAsLin3) is comprehensive R package for efficiently determining multivariable association between clinical metadata and microbial meta-omics features. MaAsLin3 relies on general linear models to accommodate most modern epidemiological study designs, including cross-sectional and longitudinal, along with a variety of filtering, normalization, and transform methods.

If you use the MaAsLin3 software, please cite our manuscript: 

William A. Nickols, Jacob T. Nearing, Kelsey N. Thompson, Jiaxian Shen, Curtis Huttenhower 
MaAsLin 3: Refining and extending generalized multivariate linear models for meta-omic association discovery. (In progress).

Check out the [MaAsLin 3 tutorial](https://github.com/biobakery/biobakery/wiki/MaAsLin3) for an overview of analysis options.

If you have questions, please direct it to :   
[MaAsLin3 Forum](https://forum.biobakery.org/c/Downstream-analysis-and-statistics/MaAsLin)    

--------------------------------------------

## Contents ##
- [MaAsLin3 User Manual](#maaslin3-user-manual)
  - [Contents](#contents)
  - [Description](#description)
  - [Requirements](#requirements)
  - [Installation](#installation)
  - [How to Run](#how-to-run)
    - [Input Files](#input-files)
    - [Output Files](#output-files)
    - [Run a Demo](#run-a-demo)
      - [Command line](#command-line)
      - [In R](#in-r)
        - [Session Info](#session-info)
    - [Options](#options)
  - [Troubleshooting](#troubleshooting)

## Description ##

MaAsLin 3 finds abundance and prevalence associations between microbiome meta-omics features and complex metadata in population-scale epidemiological studies. The software includes multiple analysis methods (including support for multiple covariates, repeated measures, and ordered predictors), filtering, normalization, and transform options to customize analysis for your specific study.

## Requirements ##

MaAsLin3 is an R package that can be run on the command line or as an R function.

## Installation ##

#### Install using GitHub and devtools

```{r, eval=FALSE}
if (!require('devtools', character.only = TRUE)) {
  install.packages('devtools')
}

library("devtools")
install_github("biobakery/MaAsLin3")
library("maaslin3")
```

## How to Run ##

MaAsLin3 can be run from the command line or as an R function. Both methods require the same arguments, have the same options, and use the same default settings.

### Input Data ###

MaAsLin3 requires two input files.

A data frame of feature abundances or read counts or a filepath to a tab-delimited file with abundances. It should be formatted with features as columns and samples as rows (or the transpose). The column and row names should be the feature names and sample names respectively.

1. Feature abundance data frame
    * Formatted with features as columns and samples as rows.
    * The transpose of this format is also okay.
    * Possible features in this file include taxonomy or genes.
    * This can be a filepath to a tab-delimited file.
2. Metadata data frame
    * Formatted with variables as columns and samples as rows.
    * The transpose of this format is also okay.
    * Possible metadata in this file include gender or age.
    * This can be a filepath to a tab-delimited file.

The data file can contain samples not included in the metadata file
(along with the reverse case). For both cases, those samples not 
included in both files will be removed from the analysis. 
Also the samples do not need to be in the same order in the two files.

NOTE: If running MaAsLin3 from the command line, the data and metadata inputs can be filepaths.

### Output Files ###

MaAsLin3 generates two types of output files: data and visualization.

1. Data output files
    * ``all_results.tsv``
        * This includes the same data as the data.frame returned.
        * This file contains all results ordered by increasing q-value.
        * The first columns are the metadata and feature names.
        * The next two columns are the value and coefficient from the model.
        * The next column is the standard deviation from the model.
        * The ``N`` column is the total number of data points.
        * The ``N.not.zero`` column is the total of non-zero data points.
        * The pvalue from the calculation is the second to last column.
        * The qvalue is computed with `p.adjust` with the correction method.
    * ``significant_results.tsv``
        * This file is a subset of the results in the first file.
        * It only includes associations with q-values <= to the threshold.
    * ``features```
        * This folder includes the filtered, normalized, and transformed versions of the input feature table.
        * These steps are performed sequentially in the above order.
        * If an option is set such that a step does not change the data, the resulting table will still be output.
    * ``models.rds``
        * This file contains a list with every model fit object.
        * It will only be generated if save_models is set to TRUE.
    * ``residuals.rds``
        * This file contains a data frame with residuals for each feature.
    * ``fitted.rds``
        * This file contains a data frame with fitted values for each feature.
    * ``ranef.rds``
        * This file contains a data frame with extracted random effects for each feature (when random effects are specified).
    * ``MaAsLin3.log``
        * This file contains all log information for the run.
        * It includes all settings, warnings, errors, and steps run.
2. Visualization output files
    * ``heatmap.pdf``
        * This file contains a heatmap of the significant associations.
    * ``[a-z/0-9]+.pdf``
        * A plot is generated for each significant association.
        * Scatter plots are used for continuous metadata.
        * Box plots are for categorical data.
        * Data points plotted are after filtering but prior to normalization and transform.

### Run a Demo ###

Example input files can be found in the ``inst/extdata`` folder 
of the MaAsLin3 source. The files provided were generated from
the HMP2 data which can be downloaded from https://ibdmdb.org/ .

``HMP2_taxonomy.tsv``: is a tab-demilited file with species as columns and samples as rows. It is a subset of the taxonomy file so it just includes the species abundances for all samples.

``HMP2_metadata.tsv``: is a tab-delimited file with samples as rows and metadata as columns. It is a subset of the metadata file so that it just includes some of the fields.


#### Command line ####

``./R/maaslin3.R inst/extdata/HMP2_taxonomy.tsv inst/extdata/HMP2_metadata.tsv command_line_output --formula='~ diagnosis + dysbiosisUC + dysbiosisCD + antibiotics + age + reads_filtered + (1 | subject)'``

* Make sure to provide the full path to the MaAsLin3 executable (i.e. `./R/MaAsLin3.R`).
* In the demo command:
    * ``inst/extdata/HMP2_taxonomy.tsv`` is the path to your data (or features) file
    * ``inst/extdata/HMP2_metadata.tsv`` is the path to your metadata file
    * ``command_line_output`` is the path to the folder to write the output


#### In R ####

```{r}
library(maaslin3)

taxa_table_name <- system.file("extdata", "HMP2_taxonomy.tsv", package = "maaslin3")
taxa_table <- read.csv(taxa_table_name, sep = '\t')
rownames(taxa_table) <- taxa_table$ID; taxa_table$ID <- NULL

# Read metadata table
metadata_name <- system.file("extdata", "HMP2_metadata.tsv", package = "maaslin3")
metadata <- read.csv(metadata_name, sep = '\t')
rownames(metadata) <- metadata$ID; metadata$ID <- NULL

# Prepare parameter lists 
param_list <- list(input_data = taxa_table, 
                   input_metadata = metadata, 
                   output = 'output', 
                   normalization = 'TSS', 
                   transform = 'LOG', 
                   formula = '~ diagnosis + dysbiosisUC + dysbiosisCD + antibiotics + age + reads_filtered + (1 | subject)', 
                   save_scatter = FALSE, 
                   save_models = FALSE, 
                   plot_summary_plot = FALSE, 
                   plot_scatter = FALSE, 
                   max_significance = 0.1, 
                   augment = TRUE, 
                   median_comparison_abundance = TRUE, 
                   median_comparison_prevalence = FALSE, 
                   cores=1)
```

##### Session Info #####

Session info from running the demo in R can be displayed with the following command.

```{r}
sessionInfo()
```

### Options ###

Run MaAsLin3 help to print a list of the options and the default settings.


$ MaAsLin3.R --help
Usage: ./R/MaAsLin3.R [options] <data.tsv> <metadata.tsv> <output_folder>


Options:
    -h, --help
        Show this help message and exit

    -a MIN_ABUNDANCE, --min_abundance=MIN_ABUNDANCE
        The minimum abundance for each feature [ Default: 0 ]

    -p MIN_PREVALENCE, --min_prevalence=MIN_PREVALENCE
        The minimum percent of samples for which a feature 
        is detected at minimum abundance [ Default: 0.1 ]

    -b MIN_VARIANCE, --min_variance=MIN_VARIANCE
        Keep features with variance greater than [ Default: 0.0 ]

    -s MAX_SIGNIFICANCE, --max_significance=MAX_SIGNIFICANCE
        The q-value threshold for significance [ Default: 0.25 ]

    -n NORMALIZATION, --normalization=NORMALIZATION
        The normalization method to apply [ Default: TSS ]
        [ Choices: TSS, CLR, CSS, NONE, TMM ]

    -t TRANSFORM, --transform=TRANSFORM
        The transform to apply [ Default: LOG ]
        [ Choices: LOG, LOGIT, AST, NONE ]

    -m ANALYSIS_METHOD, --analysis_method=ANALYSIS_METHOD
        The analysis method to apply [ Default: LM ]
        [ Choices: LM, CPLM, NEGBIN, ZINB ]

    -r RANDOM_EFFECTS, --random_effects=RANDOM_EFFECTS
        The random effects for the model, comma-delimited
        for multiple effects [ Default: none ]

    -f FIXED_EFFECTS, --fixed_effects=FIXED_EFFECTS
        The fixed effects for the model, comma-delimited
        for multiple effects [ Default: all ]

    -c CORRECTION, --correction=CORRECTION
        The correction method for computing the 
        q-value [ Default: BH ]

    -z STANDARDIZE, --standardize=STANDARDIZE
        Apply z-score so continuous metadata are 
        on the same scale [ Default: TRUE ]

    -l PLOT_HEATMAP, --plot_heatmap=PLOT_HEATMAP
        Generate a heatmap for the significant 
        associations [ Default: TRUE ]

    -i HEATMAP_FIRST_N, --heatmap_first_n=HEATMAP_FIRST_N
        In heatmap, plot top N features with significant 
        associations [ Default: TRUE ]

    -o PLOT_SCATTER, --plot_scatter=PLOT_SCATTER
        Generate scatter plots for the significant
        associations [ Default: TRUE ]
        
    -g MAX_PNGS, --max_pngs=MAX_PNGS
        The maximum number of scatter plots for signficant associations 
        to save as png files [ Default: 10 ]
    
    -O SAVE_SCATTER, --save_scatter=SAVE_SCATTER
        Save all scatter plot ggplot objects
        to an RData file [ Default: FALSE ]

    -e CORES, --cores=CORES
        The number of R processes to run in parallel
        [ Default: 1 ]
        
    -j SAVE_MODELS --save_models=SAVE_MODELS
        Return the full model outputs and save to an RData file
        [ Default: FALSE ]

    -d REFERENCE, --reference=REFERENCE
        The factor to use as a reference level for a categorical variable 
        provided as a string of 'variable,reference', semi-colon delimited for 
        multiple variables. Not required if metadata is passed as a factor or 
        for variables with less than two levels but can be set regardless.
        [ Default: NA ] 

## Troubleshooting ##

1. Question: When I run from the command line I see the error ``MaAsLin3.R: command not found``. How do I fix this? 
    * Answer: Provide the full path to the executable when running MaAsLin3.R.
2. Question: When I run as a function I see the error ``Error in library(MaAsLin3): there is no package called 'MaAsLin3'``. How do I fix this? 
    * Answer: Install the R package and then try loading the library again.
3. Question: When I try to install the R package I see errors about dependencies not being installed. Why is this?
    * Answer: Installing the R package will not automatically install the packages MaAsLin3 requires. Please install the dependencies and then install the MaAsLin3 R package.