---
title: "MaAsLin 3 User Manual"
author:
 - name: William Nickols
   email: willnickols@g.harvard.edu
 - name: Jacob Nearing
   email: nearing@broadinstitute.org
 - name: Sagun Maharjan
   email: smaharjan@hsph.harvard.edu
output: html_document
vignette: >
    %\VignetteIndexEntry{Manual}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

## Introduction ##

MaAsLin 3 is the next generation of MaAsLin (Microbiome Multivariable Association with Linear Models).

[MaAsLin3](http://huttenhower.sph.harvard.edu/MaAsLin3) is comprehensive R package for efficiently determining multivariable association between clinical metadata and microbial meta-omics features. MaAsLin3 relies on general linear models to accommodate most modern epidemiological study designs, including cross-sectional and longitudinal, along with a variety of filtering, normalization, and transform methods.

If you use the MaAsLin3 software, please cite our manuscript: 

William A. Nickols, Jacob T. Nearing, Kelsey N. Thompson, Jiaxian Shen, Curtis Huttenhower 
MaAsLin 3: Refining and extending generalized multivariate linear models for meta-omic association discovery. (In progress).

Check out the [MaAsLin 3 tutorial](https://github.com/biobakery/biobakery/wiki/MaAsLin3) for an overview of analysis options.

If you have questions, please direct it to: [MaAsLin 3 Forum](https://forum.biobakery.org/c/Downstream-analysis-and-statistics/MaAsLin). 

--------------------------------------------

## Contents ##
- [Introduction](#introduction)
- [Contents](#contents)
- [Description](#description)
- [Requirements](#requirements)
- [Installation](#installation)
- [How to run](#how-to-run)
  - [Input data](#input-data)
  - [Output files](#output-files)
  - [Run a demo](#run-a-demo)
    - [In R](#in-r)
      - [Plot revisions](#plot-revisions)
      - [Session info](#session-info)
    - [Command line](#command-line)
  - [Options](#options)
    - [Required parameters](#required-parameters)
    - [Model formula](#model-formula)
    - [Analysis options](#analysis-options)
    - [Compositionality corrections](#compositionality-corrections)
    - [Plotting parameters](#plotting-parameters)
    - [Technical parameters](#technical-parameters)
- [Troubleshooting](#troubleshooting)

## Description ##

MaAsLin 3 finds abundance and prevalence associations between microbiome meta-omics features and complex metadata in population-scale epidemiological studies. The software includes multiple analysis methods (including support for multiple covariates, repeated measures, and ordered predictors), filtering, normalization, and transform options to customize analysis for your specific study.

## Requirements ##

MaAsLin3 is an R package that can be run on the command line or as an R function.

## Installation ##

#### Install using GitHub and devtools

```{r, eval=FALSE}
if (!require('devtools', character.only = TRUE)) {
  install.packages('devtools')
}

library("devtools")
install_github("biobakery/MaAsLin3")
library("maaslin3")
```

## How to run ##

MaAsLin3 can be run from the command line or as an R function. Both methods require the same arguments, have the same options, and use the same default settings.

### Input data ###

MaAsLin3 requires two input files.

1. Feature abundance data frame
    * Formatted with features as columns and samples as rows.
    * The transpose of this format is also okay.
    * Possible features in this file include taxonomy or genes.
    * This can be a filepath to a tab-delimited file.
2. Metadata data frame
    * Formatted with variables as columns and samples as rows.
    * The transpose of this format is also okay.
    * Possible metadata in this file include gender or age.
    * This can be a filepath to a tab-delimited file.

The data file can contain samples not included in the metadata file
(along with the reverse case). For both cases, those samples not 
included in both files will be removed from the analysis. 
Also the samples do not need to be in the same order in the two files.

NOTE: If running MaAsLin3 from the command line, the data and metadata inputs can be filepaths.

### Output files ###

MaAsLin3 generates two types of output files: data and visualization.

1. Data output files
    * ``all_results.tsv``
        * This file contains all results ordered by increasing q-value.
        * Columns 1 and 2 are the feature and metadata names.
        * Columns 3 and 4 are the value of the metadata and variable name from the model.
        * Columns 5 and 6 are the fit coefficient and standard error from the model.
        * Column 7 is the p-value of the individual association.
        * Column 8 lists any errors from the model fitting.
        * Column 9 is the q-value of the individual association.
        * Column 10 species whether the association is abundance or prevalence.
        * Column 11 is the total number of data points.
        * Column 12 is the total number of non-zero data points.
        * Columns 13 and 14 are the p-value and q-value of the joint prevalence and abundance association.
    * ``significant_results.tsv``
        * This file is a subset of the results in the first file.
        * It only includes associations with q-values <= to the threshold.
    * ``features``
        * This folder includes the filtered, normalized, and transformed versions of the input feature table.
        * These steps are performed sequentially in the above order.
        * If an option is set such that a step does not change the data, the resulting table will still be output.
    * ``models_LM.rds`` and ``models_logistic.rds``
        * This file contains a list with every model fit object (`LM` for linear models, `logistic` for logistic models).
        * It will only be generated if `save_models` is set to TRUE.
    * ``residuals_LM.rds`` and ``residuals_logstic.rds``
        * This file contains a data frame with residuals for each feature.
    * ``fitted_LM.rds`` and ``fitted_logistic.rds``
        * This file contains a data frame with fitted values for each feature.
    * ``ranef_LM.rds`` and ``ranef_logistic.rds``
        * This file contains a data frame with extracted random effects for each feature (when random effects are specified).
    * ``maaslin3.log``
        * This file contains all log information for the run.
        * It includes all settings, warnings, errors, and steps run.
2. Visualization output files
    * ``summary_plot.pdf``
        * This file contains a combined coefficient plot and heatmap of the most significant associations.
    * ``association_plots/[metadatum]_[feature]_[association].pdf``
        * A plot is generated for each significant association up to `max_pngs`.
        * Scatter plots are used for continuous metadata abundance associations.
        * Box plots are used for categorical data abundance associations.
        * Box plots are used for continuous data prevalence associations.
        * Grids are used for categorical data prevalence associations.
        * Data points plotted are after filtering, normalization, and transformation so that the scale in the plot is the scale that was used in fitting.

### Run a demo ###

Example input files can be found in the ``inst/extdata`` folder 
of the MaAsLin 3 source. The files provided were generated from
the HMP2 data which can be downloaded from https://ibdmdb.org/ .

* ``HMP2_taxonomy.tsv``: is a tab-demilited file with species as columns and samples as rows. It is a subset of the taxonomy file so it just includes some of the species abundances.

* ``HMP2_metadata.tsv``: is a tab-delimited file with samples as rows and metadata as columns. It is a subset of the metadata file so that it just includes some of the fields.

#### In R ####

The following code identifies associations between patient metadata and sample species in the Human Microbiome Project 2 cohort. There are a few things to highlight:

* Because MaAsLin 3 models prevalence associations, the number of reads should be included as a covariate since deeper sequencing will enable better detection of features.
* We set `median_comparison_abundance = TRUE` but `median_comparison_prevalence = FALSE` since we want to compare the abundance coefficients against their medians to account for compositionality, but we just want to compare the prevalence coefficients against 0. If we instead wanted to see whether a feature's prevalence coefficient was significantly different from the median, we could set `median_comparison_prevalence = TRUE`.
* When warnings or errors are thrown during the fitting process, they are recorded in the `error` column of the outputs. Often, these indicate model fitting failures or poor fits that should not be trusted, but sometimes the warnings can be benign, and the model fit might still be reasonable. Users should check associations of interest if they produce errors.
* In the resulting `figures/` folder, it is a good idea to check the association plots for the most significant associations. In particular, significant categorical prevalence associations should have a reasonable number of samples in each grid box (e.g., >10 present and >10 absent for each significant category).

```{r, eval=F}
library(maaslin3)

# Read abundance table
taxa_table_name <- system.file("extdata", "HMP2_taxonomy.tsv", package = "maaslin3")
taxa_table <- read.csv(taxa_table_name, sep = '\t')

# Read metadata table
metadata_name <- system.file("extdata", "HMP2_metadata.tsv", package = "maaslin3")
metadata <- read.csv(metadata_name, sep = '\t')

metadata$diagnosis <- 
  factor(metadata$diagnosis, levels = c('nonIBD', 'UC', 'CD'))
metadata$dysbiosis_state <- 
  factor(metadata$dysbiosis_state, levels = c('none', 'dysbiosis_UC', 'dysbiosis_CD'))
metadata$antibiotics <- 
  factor(metadata$antibiotics, levels = c('No', 'Yes'))

# Prepare parameter lists 
param_list <- list(input_data = taxa_table, 
                   input_metadata = metadata, 
                   output = 'output', 
                   normalization = 'TSS', 
                   transform = 'LOG', 
                   formula = '~ diagnosis + dysbiosis_state + antibiotics + age + reads',
                   save_models = FALSE, 
                   plot_summary_plot = TRUE, 
                   max_significance = 0.1, 
                   augment = TRUE, 
                   median_comparison_abundance = TRUE, 
                   median_comparison_prevalence = FALSE, 
                   cores=1)

# Run MaAsLin 3
fit_out <- maaslin3(param_list)
```

The `maaslin3` wrapper function runs all steps at once, but equivalent results would be produced from running one step at a time:

```{r, eval=F}
param_list <- maaslin_log_arguments(param_list)
params_and_data <- maaslin_read_data(param_list)
params_and_data <- maaslin_reorder_data(params_and_data)
params_and_data_and_formula <- maaslin_check_formula(params_and_data)
params_and_data_and_formula <- maaslin_filter_and_standardize(params_and_data_and_formula)
params_and_data_and_formula <- maaslin_normalize(params_and_data_and_formula)
params_and_data_and_formula <- maaslin_transform(params_and_data_and_formula)
fit_out <- maaslin_fit(params_and_data_and_formula)
maaslin_write_results(fit_out)
invisible(maaslin_plot_results(fit_out)) # invisible to avoid dumping plots into Rmd
```

#### Plot revisions ####

Once `maaslin3` has been run once, `maaslin_plot_results_from_output` can be run with the same `param_list` as the original `maaslin3` run to (re-)create the plots. This allows the user to plot the associations even without having the R object returned by `maaslin_fit` or `maaslin3` (e.g., if fitting models through the command line). Also, the user can change which variables are specified as `coef_plot_vars` and `heatmap_vars` to change which associations are plotted. Finally, a user might fit a model with a simple variable name that will be robust to formula parsing (e.g., `dysbiosis_state`) or full taxa names but want the plots to contain more professional or readable versions of the names. In this case, the user can edit the `all_results.tsv` file to contain the revised names and then rerun only the plotting. When recreating only the summary plot (i.e., `plot_associations = FALSE`), only names in the `all_results.tsv` file need to be edited. However, when recreating the individual association plots (i.e., `plot_associations = TRUE`), names should be consistently edited across the `all_results.tsv` file and the `input_metadata` and `input_data` parameters of the `param_list`.

##### Session info #####

Session info from running the demo in R can be displayed with the following command.

```{r}
sessionInfo()
```

#### Command line ####

```{r, engine = 'bash', eval = FALSE}
./R/maaslin3.R inst/extdata/HMP2_taxonomy.tsv inst/extdata/HMP2_metadata.tsv command_line_output  --formula='~ diagnosis + dysbiosis_state + antibiotics + age + reads' --reference='diagnosis,nonIBD;dysbiosis_state,none;antibiotics,No'
```

* Make sure to provide the full path to the MaAsLin3 executable (i.e. `./R/maaslin3.R`).
* In the demo command:
    * ``inst/extdata/HMP2_taxonomy.tsv`` is the path to your data (or features) file
    * ``inst/extdata/HMP2_metadata.tsv`` is the path to your metadata file
    * ``command_line_output`` is the path to the folder to write the output

### Options ###

From the command line, the following command will print the list of MaAsLin 3 options and default settings:

```$ ./R/maaslin3.R --help```

When running MaAsLin 3 in R, the manual page for each function (e.g., `?maaslin3`) will show the available options and default settings. For both, the options and settings are as follows:

#### Required parameters ####

* `input_data`: A data frame of feature abundances or read counts or a filepath to a tab-delimited file with abundances. It should be formatted with features as columns and samples as rows (or the transpose). The column and row names should be the feature names and sample names respectively.
* `input_metadata`: A data frame of per-sample metadata or a filepath to a tab-delimited file with metadata. It should be formatted with variables as columns and samples as rows (or the transpose). The column and row names should be the variable names and sample names respectively.
* `output`: The output folder to write results.

#### Model formula ####

* `formula`: A formula in `lme4` format. Random effects, interactions, and functions of the metadata can be included (note that these functions will be applied after standardization if `standardize = TRUE`). Group and ordered variables can be specified as: `group(grouping_variable)` and `ordered(ordered_variable)`. The other variable options below will not be considered if a formula is set.
* `fixed_effects`: A vector of variable names to be included as fixed effects.
* `reference`: For a variable with more than two levels supplied with `fixed_effects`, the factor to use as a reference provided as a string of 'variable,reference' semi-colon delimited for multiple variables.
* `random_effects`: A vector of variable names to be included as random intercepts.
* `group_effects`: A factored categorical variable to be included for group testing. An ANOVA-style test will be performed to assess whether any of the variable's levels are significant, and no coefficients or individual p-values will be returned.
* `ordered_effects`: A factored categorical variable to be included. Consecutive levels will be tested for significance against each other, and the resulting associations will correspond to effect sizes, standard errors, and significances of each level versus the previous.

#### Analysis options ####

* `min_abundance` (default `0`): Features with abundances of at least `min_abundance` in `min_prevalence` of the samples will be included for analysis. The threshold is applied before normalization and transformation. The options `min_abundance` and `min_prevalence` should usually be 0 since MaAsLin 3 is designed to model sparsity.
* `min_prevalence` (default `0`): See above.
* `zero_threshold` (default `0`): Abundances less than or equal to `zero_threshold` will be treated as zeros. This is primarily to be used when the abundance table has likely low-abundance false positives.
* `min_variance` (default `0`): Features with abundance variances less than or equal to `min_variance` will be dropped. This is primarily used for dropping features that are entirely zero.
* `max_significance` (default `0.1`): The FDR corrected q-value threshold for significance used in selecting which associations to write as significant and to plot.
* `normalization` (default `TSS`): The normalization to apply to the features before transformation and analysis. The option `TSS` is recommended, but `CLR`, `CSS`, `NONE`, and `TMM` can also be used.
* `transform` (default `LOG`): The transformation to apply to the features after normalization and before analysis. The option `LOG` is recommended, but `LOGIT`, `AST`, and `NONE` can also be used.
* `correction` (default `BH`): The correction to obtain FDR-corrected q-values from raw p-values. Any valid options for `p.adjust` can be used.
* `standardize` (default `TRUE`): Whether to apply z-scores to continuous metadata variables so they are on the same scale. This is recommended in order to compare coefficients across metadata variables, but note that functions of the metadata specified in the `formula` will apply after standardization.

#### Compositionality corrections ####

* `unscaled_abundance`: A data frame with a single column of absolute abundances or a filepath to such a tab-delimited file. The row names should match the names of the samples in `input_data` and `input_metadata`. When using spike-ins, the single column should have the same name as one of the features in `input_data`, and the `unscaled_abundance` should correspond to the absolute quantity of the spike-in. For example, if the same spike-in quantity is used in each sample, the entire column can be set to 1. When using total abundance scaling, the single column should have the name 'total', and the `unscaled_abundance` should correspond to the total abundance of each sample. In both cases, `median_comparison_abundance` should be set to `FALSE` since the spike-in or total abundance normalization accounts for compositionality.

  Alternatively, if the `input_data` abundances have already been scaled to be absolute abundances, the user should set `normalization = NONE` and `median_comparison_abundance = FALSE` and not include anything for `unscaled_abundance`. 
* `median_comparison_abundance` (default `TRUE`): Test abundance coefficients against a null value corresponding to the median coefficient for a metadata variable across the features. This is recommended for relative abundance data but should not be used for absolute abundance data.
* `median_comparison_prevalence` (default `FALSE`): Test prevalence coefficients against a null value corresponding to the median coefficient for a metadata variable across the features. This is only recommended if the analyst is interested in how feature prevalence associations compare to each other or if there is likely strong compositionality-induced sparsity.
* `median_comparison_abundance_threshold` (default `0.25`): Coefficients within `median_comparison_abundance_threshold` of the median association will automatically be counted as insignificant (p-value set to 1) since they likely represent compositionality-induced associations. This threshold will be divided by the metadata variable's standard deviation if the metadatum is continuous to ensure the threshold applies to the right scale.
* `median_comparison_prevalence_threshold` (default `0.25`): Same as `median_comparison_abundance_threshold` but applied to the prevalence associations.
* `augment` (default `TRUE`): Add extra lowly-weighted 0s and 1s to avoid linear separability.

#### Plotting parameters ####

* `plot_summary_plot` (default `TRUE`): Generate a summary plot of significant associations.
* `summary_plot_first_n` (default `25`): Include the top `summary_plot_first_n` features with significant associations.
* `coef_plot_vars`: Vector of variable names to be used in the coefficient plot section of the summary plot. Continuous variables should match the metadata column name, and categorical variables should be of the form `"[variable] [level]"`.

  By default, the (up to) two metadata variables with the most significant associations will be plotted in the coefficient plot, and the rest will be plotted in the heatmap. Because predicting the output variable names can be tricky, it is recommended to first run `maaslin3` without setting `coef_plot_vars` or `heatmap_vars`, look at the names of the variables in the summary plot, and then rerun with `maaslin_plot_results_from_output` after updating `coef_plot_vars` and `heatmap_vars` in `param_list` with the desired variables.
* `heatmap_vars`: Vector of variable names to be used in the heatmap section of the summary plot. Continuous variables should match the metadata column name, and categorical variables should be of the form `"[variable] [level]"`.
* `plot_associations` (default `TRUE`): Whether to generate plots for significant associations.
* `max_pngs` (default `30`): The top `max_pngs` associations will be plotted.

#### Technical parameters ####

* `cores` (default `1`): How many cores to use when fitting models. (Using multiple cores will likely be faster only for large datasets or complex models.)
* `save_models` (default `FALSE`): Whether to return the fit models and save them to an RData file.
		
## Troubleshooting ##

1. Question: When I run from the command line I see the error ``maaslin3.R: command not found``. How do I fix this? 
    * Answer: Provide the full path to the executable when running maaslin3.R
2. Question: When I run as a function I see the error ``Error in library(maaslin3): there is no package called 'maaslin3'``. How do I fix this? 
    * Answer: Install the R package and then try loading the library again.
